RPP - Connect the duplicate payment prevention service (Verification)

htdat opened the pull request:
Fixes #7411

## Changes proposed in this Pull Request

- Create new `DuplicatePaymentPreventionService` class in `src` and keep most of logics from the legacy one.
- Hooks: remove the hook from old legacy service, and register hooks from the new `src` service. 
- Introduce only one new state `DuplicateOrderDetectedState` when a duplicate order is detected.
- Introduce `Intent_Status::AUTHORIZED_STATUSES` and `is_authorized()` for intent abstraction.
- Add `SessionService` in `src` with the purpose of easily handling session data (sometimes, it's null while other times, it's really the expected instance of `WC_Session`). This will also be useful when other features will use it, such as Fraud relevant features.
- Add tests, cover 100% for `src` files.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions 

- At this stage, we can only test for: 
    - type of card: basic. 3DS is not yet supported: though #7471 is merged but it's still not possible to finish checkout with a 3DS card - see more details p1698122299758599-slack-C05997J5PCG
    - checkout page type: both classic and block.
    - method: both new and saved method.
    - gateway type: does not matter as we only have one type for the new payment process. 

- Ensure that you're using the latest versions of Dev Tools and enable **New Payment Process Factors**.
- Ideally, add Xdebug breakpoints in the new code in InitialState.php and the new service php file to confirm new code really executed. 

### For testing duplicate orders 

Basically, follow test instruction and result of this PR https://github.com/Automattic/woocommerce-payments/pull/5247 with a note: 
- Use this updated snippet https://gist.github.com/htdat/757b1375146584e63ebad80c5d02f1a1 (rather than the one in PR 5247) 

### For testing duplicate payments 

Basically, follow test instruction and result of this PR https://github.com/Automattic/woocommerce-payments/pull/5346 with a note: 

- For the diff, please use this new one https://gist.github.com/htdat/14715ca445cea3d4d704d5d1a924b25f
- For the snippet, it's OK to use the one mentioned in PR 5346 https://gist.github.com/htdat/6a52eec1be77b8442ac05401908a3627


<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

## Remaining TODO(s) for this PR: 

N/A

## Finished TODOs for this PR

- [x] Consider port this change https://github.com/Automattic/woocommerce-payments/pull/7441 - We may not need this as the fix for PR 7441 is for legacy UPE, and split UPE (not dUPE), which are going to be deprecated. 
- [x] Use `Logger` as dependency per https://github.com/Automattic/woocommerce-payments/issues/7428
- [x] Consider including this refactoring https://github.com/Automattic/woocommerce-payments/issues/6712
- [x] Register hooks but avoid duplicate hooks https://github.com/Automattic/woocommerce-payments/pull/7450#discussion_r1356675850
- [x] Update to use a post-procesing state https://github.com/Automattic/woocommerce-payments/pull/7450#discussion_r1363170224

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] NO NEED Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

htdat committed:
Copy and register Duplicate_Payment_Prevention_Service to src

htdat committed:
Remove WC_Payment_Gateway_WCPay dependency, update OrderService to private

htdat committed:
Move const SUCCESSFUL_INTENT_STATUS in gateway to Intent_Status data class

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7450` or branch name `rpp/7411-duplicate-prevention` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=rpp/7411-duplicate-prevention) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f273a60e92c46d6f04278f2d334c33031f74ff31
- Build time: 2023-10-26 08:58:51 UTC

_Note: the build is updated when a new commit is pushed to this PR._

RadoslavGeorgiev commented on includes/constants/class-intent-status.php:
```diff
@@ -26,4 +26,15 @@ class Intent_Status extends Base_Constant {
+	 *
+	 * @type array
+	 */
+	const SUCCESSFUL_STATUSES = [
```
@htdat now that we have the abstract intention, what are your thoughts on adding an `is_successful()` method to it, and using it to check the status there? By doing so, we would not need this constant, and less code would be repeated.

htdat replied:
This is a good point too. I will implement that one.

htdat replied:
@RadoslavGeorgiev - good idea, however, I think, we might think more about the method name. `is_successful()` can be confusing with "this intent status is `succeeded`". What if we come with something like `is_authorized`? Otherwise, the plural noun `statuses` forces other developers to double check what's they are.

RadoslavGeorgiev replied:
Yes, that makes sense to me. 👍

htdat replied:
Fixed 2a3d46285c3744ba9440248394da0bcdb32886bb

RadoslavGeorgiev commented on src/Internal/Service/DuplicatePaymentPreventionService.php:
```diff
@@ -0,0 +1,220 @@
+	 *
+	 * @param OrderService $order_service The order service instance.
+	 */
+	public function init( OrderService $order_service ) {
```
1. The order service should be provided in the class constructor.
2. `init` should be `init_hooks`, and we can add it in `class WC_Payments` for now.

htdat replied:
😀 Good point to notice. Anyway, I am still in the progress of working on this one. I will take care of this one.

htdat replied:
Updated the code but there is a new concern
https://github.com/Automattic/woocommerce-payments/pull/7450/files#r1356675850

htdat committed:
Replace methods for order object with respective OrderService methods

htdat committed:
Replace init with __construct for DPPS.

htdat commented on src/Internal/Service/DuplicatePaymentPreventionService.php:
```diff
@@ -0,0 +1,237 @@
+	 *
+	 * @return void
+	 */
+	public function init_hooks(): void {
```
Not yet register this `init_hook` in `WC_Payments` as that can make this action `clear_session_processing_order_after_landing_order_received_page` is registered twice. 

Still pending to think how to approach it.

htdat replied:
@RadoslavGeorgiev 

> I'm curious where `init_hooks` of the DPPS will be called. WDYT about https://github.com/Automattic/woocommerce-payments/pull/7479 (Hooks Registry)?

I like the concept of `Hooks Registry` for new `src` folder. But it's still not getting out of how to solve potential duplicate hook registers in this specific case.

RadoslavGeorgiev replied:
Do you see any reason to keep the hooked methods in the legacy service? AFAICT we can remove them, and have a single hook?

htdat replied:
No reason to keep it! If we can guarantee that the hook is always loaded, so I think we're all good. I will dive into this idea, and see how it looks like.

htdat replied:
Remove the hook from the legacy service, and register the hook from the new `src` service - 95ed11be21ad5cd011ebb4f4c4c91151a7e92369

I am going to mark this as resolved. Please re-open if needed.

htdat committed:
Update WC()->session as a dependency

htdat committed:
Update is_order_received_page and global $wp to call via LegacyProxy

htdat committed:
Change to use WC_Session to use translation layer SessionService

Ref p1697184565184799/1697109799.887649-slack-C05997J5PCG

htdat committed:
Handling previously paid orders with TODOs

htdat commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -808,6 +810,14 @@ public function new_process_payment( WC_Order $order ) {
+		if ( $state instanceof PreviousPaidOrderDetectedState ) {
+			return [
+				'result'   => 'success',
+				'redirect' => $this->get_return_url( $order ), // TODO: This should be PaymentContext::get_previous_order_id(). But how to get it as we do expose PaymentContext here at the moment.
```
@RadoslavGeorgiev - See 15dc9679b5eacf97c8e51f356eef1c50f3237ed7 - this is a tricky part. If we had a storage for PaymentContext, this would be easier.

htdat replied:
Handled here a8c478bc2e1d6c5a3f0df1dc0e5ca0e27eda65f3

htdat replied:
Resolving this now.

htdat committed:
Handle getting the order ID from PreviousPaidOrderDetectedState

htdat committed:
Merge branch 'develop' into rpp/7411-duplicate-prevention

htdat committed:
Update correct method name set_previous_paid_order_id

htdat committed:
Separate detecting duplicate order, and clean up actions

htdat committed:
Update naming - use "duplicate order"

htdat committed:
Move down the verification step of get_previous_paid_duplicate_order_id

htdat committed:
Remove maybe, use update_session_processing_order

htdat committed:
Correct return url when handle DuplicateOrderDetectedState

htdat committed:
Handling existing successful intents, change name to authorized intents

htdat committed:
Update to use is_authorized for intents

htdat commented:
I am going to add tests as well as fix Psalm errors. But it's ready to continue having more reviews. @RadoslavGeorgiev @Automattic/gamma @marcinbot

RadoslavGeorgiev commented on src/Internal/DependencyManagement/ServiceProvider/PaymentsServiceProvider.php:
```diff
@@ -47,6 +51,7 @@ class PaymentsServiceProvider extends AbstractServiceProvider {
 		ExampleService::class,
 		ExampleServiceWithDependencies::class,
 		PaymentRequestService::class,
+		DuplicatePaymentPreventionService::class,
```
Let's also add the new state here.

htdat replied:
Good point, I missed it somehow. Addressed 9b414720c672ea0fcae13b64da3a2c0166e7efbe

RadoslavGeorgiev commented on src/Internal/DependencyManagement/ServiceProvider/PaymentsServiceProvider.php:
```diff
@@ -64,12 +69,21 @@ public function register(): void {
 
 		$container->addShared( PaymentRequestService::class );
 
+		$container->addShared( PaymentRequestService::class );
```
Somehow this is added twice now.

htdat replied:
Seems a leftover. Fixed 9b414720c672ea0fcae13b64da3a2c0166e7efbe

RadoslavGeorgiev commented on src/Internal/Payment/State/DuplicateOrderDetectedState.php:
```diff
@@ -0,0 +1,15 @@
+<?php
+/**
+ * Class PreviousPaidOrderDetectedState
```
Comment from another file.

htdat replied:
Good spotting! I changed the state class name in the middle, but did not update this. 
Also fixed here 9b414720c672ea0fcae13b64da3a2c0166e7efbe

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -97,6 +108,21 @@ public function process( PaymentRequest $request ) {
+		}
+		$this->dpps->update_session_processing_order( $order_id );
+
+		$authorized_intent_id = $this->dpps->get_authorized_payment_intent_attached_to_order( $order_id );
```
We can directly use the intent object here, rather than just the ID.

htdat replied:
Together with this discussion, https://github.com/Automattic/woocommerce-payments/pull/7450#discussion_r1363181063, I am switching to your suggestion approach in 9cb3e3a8e524e94342ad50a645871d600ce03e72
Note: I am aware of the potential conflict with `get_intent` and `set_intent` with Zvnonimir's PR.

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -97,6 +108,21 @@ public function process( PaymentRequest $request ) {
+		if ( null !== $authorized_intent_id ) {
+			$this->dpps->remove_session_processing_order( $order_id );
+			$context->set_authorized_intent_id( $authorized_intent_id );
+			return $this->create_state( CompletedState::class ); // TODO. This will be updated to a new post-processing state.
```
Adding a comment for the open ToDo.

RadoslavGeorgiev commented on src/Internal/Payment/PaymentContext.php:
```diff
@@ -225,6 +225,53 @@ public function get_customer_id(): ?string {
+	 *
+	 * @return string|null Payment intent ID.
+	 */
+	public function get_authorized_intent_id(): ?string {
```
Instead of using the intent ID for logic here, let's just use a boolean flag. The ID of the authorized intent will be stored in the `intent` property already. Then this method can be removed.

htdat replied:
Agreed that we would use a boolean flag here. 
<strike>But it's not clear which `intent` property you mean here?</strike>

But anway, since the intent ID is attached to the order already, we can retrieve it anytime.

htdat replied:
> But it's not clear which intent property you mean here?

Oh, I've just found out it https://github.com/Automattic/woocommerce-payments/pull/7471/files#diff-35fbbfe7a2e6ba1cab161794c92b9001d634b741162fb3f22106c3f03cfec83cR255-R266. This makes sens now.

_Update_: addressed 9cb3e3a8e524e94342ad50a645871d600ce03e72

RadoslavGeorgiev commented on src/Internal/Service/DuplicatePaymentPreventionService.php:
```diff
@@ -0,0 +1,237 @@
+	 * @param  HooksProxy     $hooks_proxy  The hooks proxy instance.
+	 * @param  LegacyProxy    $legacy_proxy  The legacy proxy instance.
+	 */
+	public function __construct( OrderService $order_service, SessionService $session_service, HooksProxy $hooks_proxy, LegacyProxy $legacy_proxy ) {
```
For some reason `SessionService` is not detected by my editor.

htdat replied:
May you reload your editor? There is no issue in my PHPStorm.
![Markup on 2023-10-18 at 11:33:46](https://github.com/Automattic/woocommerce-payments/assets/10045087/eba4cbae-5018-4c95-a1a7-7be1c9ca2515)

RadoslavGeorgiev replied:
All good now, maybe it was some one-time editor bug.

RadoslavGeorgiev commented on src/Internal/Service/OrderService.php:
```diff
@@ -254,6 +254,87 @@ public function attach_exchange_info_to_order( int $order_id, WC_Payments_API_Ch
+	 * @throws Order_Not_Found_Exception
+	 */
+	public function update_order_status_from_intent( int $order_id, WC_Payments_API_Abstract_Intention $intent ): void {
+		$this->legacy_service->update_order_status_from_intent( $this->get_order( $order_id ), $intent );
```
See the `update_order_from_successful_intent` method of this class. Let's at least here, create separate methods for separate outcomes. This one should likely make sure to update all needed meta as well.

htdat replied:
I am going to remove this method. It was a part of `check_payment_intent_attached_to_order_succeeded` for updating orders after the currenlty attached intent has one of successful ("authorized" is more correct, with our new glossary) statuses.

https://github.com/Automattic/woocommerce-payments/blob/9b414720c672ea0fcae13b64da3a2c0166e7efbe/includes/class-duplicate-payment-prevention-service.php#L117

Eventually, this action would be handed in a new post-processing state, which is currently a TODO https://github.com/Automattic/woocommerce-payments/pull/7450#discussion_r1362622441

htdat replied:
Removed here d09d1e03d7e60e3bfde8a2dc31b14abd4f88799f

RadoslavGeorgiev commented on src/Internal/Service/OrderService.php:
```diff
@@ -254,6 +254,87 @@ public function attach_exchange_info_to_order( int $order_id, WC_Payments_API_Ch
+	 * @throws Order_Not_Found_Exception
+	 */
+	public function is_paid( int $order_id ): bool {
+		return $this->get_order( $order_id )->has_status( wc_get_is_paid_statuses() );
```
`LegacyProxy`` needed here.

htdat replied:
Not sure why we also need `LegacyProxy` here while `$this->get_order()` is actually using `LegacyProxy`?

htdat replied:
Oh, sorry, I missed `wc_get_is_paid_statuses`. Going to fix it now.

htdat replied:
01e204ce9d4671655cec53453ddbb069898f58f6 would fix it.

RadoslavGeorgiev commented on src/Internal/Service/PaymentProcessingService.php:
```diff
@@ -64,11 +64,11 @@ public function process_payment( int $order_id, bool $manual_capture = false ) {
-		$completed_state = $initial_state->process( $request );
+		$request       = new PaymentRequest( $this->legacy_proxy );
+		$initial_state = $this->state_factory->create_state( InitialState::class, $context );
+		$new_state     = $initial_state->process( $request );
```
Warning: @zmaglica's PR already returns `$initial_state->process()`, there will be a conflict here.

htdat replied:
Yes, I expect that this will happen, and ready to fix the conflict.

RadoslavGeorgiev reviewed (commented):
Great job so far, @htdat!

I have a few minor comments, but the general approach looks good. A couple of extra notes:

- I'm curious where `init_hooks` of the DPPS will be called. WDYT about https://github.com/Automattic/woocommerce-payments/pull/7479?
- It's late for me, and I haven't thoroughly checked the DPPS, but I'm confident that if I have comments there, they would be nitpicks.

mdmoore committed:
Add WooPay button location option (#7431)

htdat committed:
Address reviews (simple changes)

htdat committed:
Address review - remove OrderService::update_order_status_from_intent

See https://github.com/Automattic/woocommerce-payments/pull/7450#discussion_r1363170224

htdat committed:
Update handling get_authorized_payment_intent_attached_to_order

- Change to use set_intent, and set a flag when detecting authorized_payment_intent.

htdat committed:
Address review - Use LegacyProxy for `wc_get_is_paid_statuses`

htdat committed:
Add more checks for get_previous_paid_duplicate_order_id

- Port changes from #7741

htdat committed:
Tests - fix InitialStateTest

- Due to introduce new dep

htdat committed:
Fix Psalm errors

htdat committed:
Add tests for OrderService

bborman22 committed:
Add feature flag check to early WooPay session request (#7459)

htdat committed:
Add tests for SessionService

htdat committed:
Add tests for PaymentContext

jessy-p committed:
Adding Internal Logger (#7462)

Co-authored-by: Jessy P <jessy.pappachan@automattic.com>

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -97,6 +108,22 @@ public function process( PaymentRequest $request ) {
 		// Populate further details from the order.
 		$this->populate_context_from_order();
 
+		$duplicate_order_id = $this->dpps->get_previous_paid_duplicate_order_id( $order_id );
+		if ( ! is_null( $duplicate_order_id ) ) {
+			$this->dpps->clean_up_when_detecting_duplicate_order( $duplicate_order_id, $order_id );
+			$context->set_duplicate_order_id( $duplicate_order_id );
+			return $this->create_state( DuplicateOrderDetectedState::class );
+		}
+		$this->dpps->update_session_processing_order( $order_id );
+
+		$authorized_intent = $this->dpps->get_authorized_payment_intent_attached_to_order( $order_id );
+		if ( null !== $authorized_intent ) {
+			$this->dpps->remove_session_processing_order( $order_id );
+			$context->set_intent( $authorized_intent );
+			$context->set_detected_authorized_intent();
+			return $this->create_state( CompletedState::class ); // TODO. This will be updated to a new post-processing state.
```
I wrote everything below only to realize that you are suggesting the same 🤣. I'm leaving it here for posterity.

---

This is not something I have not considered while defining the issue, and for the most part, my thinking is that we should not allow states to be "injected" into the payment process from outside.

At first, when I saw the number of lines within the conditional blocks, it seemed a bit much for me as well, but I still prefer it this way:

```php
$reason_found = $some_service->check_something();
if ( $reason_found ) {
    $some_service->perform_action();
    return $this->createState( SomeState );
}
```

I found that the `maybe_perform_some_action_elsewhere_and_return_result()` way of coding obfuscates what's going on, and makes navigating code more difficult.

I'm willing to meet somewhere in the middle: Isolating this functionality within a method, which looks like what you described, but lives in the state.

htdat replied:
Thanks for the great idea. 
I've moved these lines to private methods, and reduce the number of lines in `InitialState::process` - b75d401fc17a67c47bd04f37ba8696c9a4f023a6

htdat committed:
Move duplicate logics to private functions in InitialState

htdat committed:
Add DuplicatePaymentPreventionServiceTest

htdat committed:
Add tests for get_authorized_payment_intent_attached_to_order

htdat committed:
Replace Logger as dep, and add tests for it

htdat committed:
Fix logic

htdat committed:
Add test_get_previous_paid_duplicate_order_id

htdat committed:
Merge branch 'develop' into rpp/7411-duplicate-prevention

htdat committed:
Add PaymentContextTest::test_intent from PR 7471

htdat committed:
InitialStateTest - fix test_process due to new protected methods

htdat committed:
Add test_process_then_detected_duplicates for InitialState

htdat committed:
Add tests for process_duplicate_order and process_duplicate_payment

htdat committed:
Register the hook in the new service, remove the similar one in the legacy.

htdat committed:
Fix psalm errors

htdat committed:
Add changelog

htdat committed:
Merge branch 'develop' into rpp/7411-duplicate-prevention

htdat commented:
Hi @marcinbot  and @Automattic/gamma  👋 PR is official ready to review, after adding the PR description and tests (to cover 100% for new changes in the `src` folder).

htdat committed:
Merge branch 'develop' into rpp/7411-duplicate-prevention

# Conflicts:
#	includes/class-wc-payment-gateway-wcpay.php
#	src/Internal/DependencyManagement/ServiceProvider/PaymentsServiceProvider.php
#	src/Internal/Payment/PaymentContext.php
#	src/Internal/Payment/State/AbstractPaymentState.php
#	src/Internal/Payment/State/InitialState.php
#	src/Internal/Service/PaymentProcessingService.php
#	tests/unit/src/Internal/Payment/PaymentContextTest.php
#	tests/unit/src/Internal/Payment/State/InitialStateTest.php

htdat committed:
Fix InitialStateTest after the merge

htdat committed:
Update @todo to ProcessedState

htdat commented:
FYI. I have already merged `develop` (including this PR https://github.com/Automattic/woocommerce-payments/pull/7471), addressed conflict, and used `ProcessedState` for the final todo beafe0c1ae748a8662517e7abc441aa42c9f2236

htdat committed:
Merge branch 'develop' into rpp/7411-duplicate-prevention

RadoslavGeorgiev commented on src/Internal/Service/SessionService.php:
```diff
@@ -0,0 +1,80 @@
+use WCPay\Internal\Proxy\LegacyProxy;
+
+/**
+ * Session service. Use as a translation layer between `src` class and Woo core WC_Session class.
```
```suggestion
 * Session service. Use as a translation layer between `src` classes and Woo core WC_Session class.
```

RadoslavGeorgiev commented on src/Internal/Service/SessionService.php:
```diff
@@ -0,0 +1,80 @@
+	/**
+	 * Getter.
+	 *
+	 * @param  string     $key Session key.
```
RadoCS: 
```suggestion
	 * @param string     $key Session key.
```

RadoslavGeorgiev commented on src/Internal/Service/SessionService.php:
```diff
@@ -0,0 +1,80 @@
+	 * Setter.
+	 *
+	 * @param  string $key Session key.
+	 * @param  mixed  $value Value to set.
```
RadoCS:
```suggestion
	 * @param string $key   Session key.
	 * @param mixed  $value Value to set.
```

RadoslavGeorgiev commented on src/Internal/Service/PaymentProcessingService.php:
```diff
@@ -123,5 +123,4 @@ protected function create_payment_context( int $order_id, bool $manual_capture =
 
 		return $context;
 	}
-
```
DatCS: 💯

RadoslavGeorgiev commented on src/Internal/Service/OrderService.php:
```diff
@@ -254,6 +254,104 @@ public function attach_exchange_info_to_order( int $order_id, WC_Payments_API_Ch
+	/**
+	 * Gets currently attached intent ID of the order.
+	 *
+	 * @param  int $order_id Order ID.
```
RadoCS:
```suggestion
	 * @param int $order_id Order ID.
```

RadoslavGeorgiev commented on src/Internal/Service/OrderService.php:
```diff
@@ -254,6 +254,104 @@ public function attach_exchange_info_to_order( int $order_id, WC_Payments_API_Ch
+	/**
+	 * Gets cart hash for the given order ID.
+	 *
+	 * @param  int $order_id  ID of the order.
```
RadoCS:
```suggestion
	 * @param int $order_id  ID of the order.
```

Note: There are a few more spaces like this in the file, but I won't add comments for them.

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -169,4 +192,52 @@ protected function populate_context_from_order() {
+	/**
+	 * Detects duplicate orders, and run the necessary actions if one is detected.
+	 *
+	 * @return AbstractPaymentState|null The next state, or null if no duplicate order is detected.
```
I think that for this method, and `process_duplicate_payment`, we can declare the concrete return classes, rather than the abstract. It won't matter much outside of this class, but would still be clearer.

htdat replied:
Good point. Addressed 1f232642905d05ac1a9fe51baff6473b5c16451a

RadoslavGeorgiev commented on src/Internal/Service/DuplicatePaymentPreventionService.php:
```diff
@@ -0,0 +1,260 @@
+		}
+
+		// We only care about payment payment_intent.
+		$is_payment_intent = 'pi_' === substr( $intent_id, 0, 3 );
+		if ( ! $is_payment_intent ) {
```
Nitpick: Could we avoid the variable, and use the condition straight within `if()`?

htdat replied:
Did it c7de0d1adf9a7ac47dc05ffb44cb7d42d4b2aaa9

RadoslavGeorgiev commented on src/Internal/Service/DuplicatePaymentPreventionService.php:
```diff
@@ -0,0 +1,260 @@
+
+		$intent_meta_order_id_raw = $payment_intent->get_metadata()['order_id'] ?? '';
+		$intent_meta_order_id     = is_numeric( $intent_meta_order_id_raw ) ? intval( $intent_meta_order_id_raw ) : 0;
+		if ( $intent_meta_order_id !== $order_id ) {
```
Question: When would this ever happen? I know it was there before, but I don't remember.

I'm asking, because maybe we should kill the payment process, rather than allowing the code to proceed here?

htdat replied:
I think we do this for: 

1. security 09ffd865c635b25d3ecffc976e8070b1f6431711.
2. dealing with the UPE intent ID session caching bug. For example, an intent ID is initially attached to order A, but then it can be reused and succeeded for order B.

Point (2) seems no longer valid. 

> I'm asking, because maybe we should kill the payment process, rather than allowing the code to proceed here?

If we kill the payment process, how the current processing order can be paid? I think letting it continue here is properly.

RadoslavGeorgiev replied:
When I wrote the comment, I was considering clearing the cart, and forcing the customer into a new order. However, if we do that, it will probably make more sense as a general part of the payment process, not something deep within the DPPS. Marking this as resolved.

RadoslavGeorgiev reviewed (changes requested):
Great work, @htdat! 👏 

I essentially only have nitpicks about some spaces here and there, but code-wise this looks great. I'm about to start testing, but wanted to share those first.

Question: Should we call `remove_session_processing_order` when transitioning from `ProcessedState` to `CompletedState`?

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/SessionServiceTest.php:
```diff
@@ -0,0 +1,140 @@
+		$this->mock_legacy_proxy = $this->createMock( LegacyProxy::class );
+		$this->sut               = new SessionService( $this->mock_legacy_proxy );
+	}
+
```
RadoCS: Too many empty spaces throughout this file here.

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/DuplicatePaymentPreventionServiceTest.php:
```diff
@@ -0,0 +1,480 @@
+
+		?int $expected
+	) {
+		$this->sut = $this->getMockBuilder( DuplicatePaymentPreventionService::class )
```
Let's add an inline comment here, defining the type correctly. Right now I'm getting errors in VSCode that DPPS methods are not defined.

Same applies to every instance of `$this->sut = $this->getMockBuilder` in the file.

```suggestion
		/**
		 * @var DuplicatePaymentPreventionService|MockObject
		 */
		$this->sut = $this->getMockBuilder( DuplicatePaymentPreventionService::class )
```

htdat replied:
Good call. I've created a separated $mocked_sut for tests using SUT with mocked methods.
621dd122fe1bb87df8a7c5677fe237e915896c08

RadoslavGeorgiev replied:
Thank you, but we still need to add doc blocks. VSCode is too smart about it, picking up the concrete type from the method: 

<img width="1063" alt="Screenshot 2023-10-26 at 8 22 42" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/80a86af5-fa63-4d05-8b07-563f08a84133">

With that in mind, and given the fact that `mock_sut` is not defined within `setUp()`, I'd actually prefer to keep it as a local variable, instead of a property.

htdat replied:
> but we still need to add doc blocks. 

That's because I missed changing `sut` to `mocked_sut` for this test `test_clean_up_when_detecting_duplicate_order`. I've updated it e688822fc

I think it should be OK now. 

> given the fact that mock_sut is not defined within setUp(), I'd actually prefer to keep it as a local variable, instead of a property.

Compared with the benefit that we do not need to declare the mixed (MockObject and concrete type of the mock) type for every `$this->getMockBuilder()` call, IMO, this is acceptable.

htdat replied:
Also, `$this->mocked_sut` can help distinguish it from `$this->sut`. 

That said, I don't mind much if we change all current `$this->mocked_sut` back to `$this->sut` and declare the type of it. Let me know WDYT :D

RadoslavGeorgiev replied:
> Compared with the benefit that we do not need to declare the mixed (MockObject and concrete type of the mock) type for every $this->getMockBuilder() call, IMO, this is acceptable.

Take another look at the screenshot. I (also) didn't notice that `$this->sut` was used instead of `$this->mock_sut`, but the error is still applicable: VSCode does not care of the PHPDoc type of `$this->var`, and infers the type from the assignment earlier in the method. So we do need to define the type regardless, and with that in mind, the private property becomes useless.

htdat replied:
I think which IDE being used does matter here :D 

With the current code (declare the mix type for private property `$this->mocked_sut`), PhpStorm can infer the correct type. 
![Markup on 2023-10-26 at 14:33:19](https://github.com/Automattic/woocommerce-payments/assets/10045087/f403ef8f-94d8-4794-94f7-84eb9ab8d645)

Without any declaration, just calling `$this->getMockBuilder`, PhpStorm can also infer the correct type: 

![Markup on 2023-10-26 at 14:37:27](https://github.com/Automattic/woocommerce-payments/assets/10045087/3f685af4-141a-4440-b217-dd02741f6b35)

--- 

Anyway, I am going to give it another try with your suggestion ff25309b639e7d465e44bc86e458861020def1af

If it's still not working for VSCode properly, might you submit a commit for it? :D

RadoslavGeorgiev replied:
> I think which IDE being used does matter here :D

It does. Seems like PhpStorm uses a combination of the property DocBlock and the assignment, joining them. In VSCode, the assignment takes priority, and the property is ignored.

Anyway, thank you for this! It works for me now.

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/DuplicatePaymentPreventionServiceTest.php:
```diff
@@ -0,0 +1,480 @@
+			$this->mock_legacy_proxy,
+		];
+
+		$this->sut = new DuplicatePaymentPreventionService( ... $this->deps );
```
Cool trick with the dependencies!

RadoslavGeorgiev reviewed (changes requested):
A couple of comments regarding tests.

htdat committed:
Add mocked_suit in DuplicatePaymentPreventionServiceTest

htdat committed:
Simplify detecting payment intent

htdat committed:
Multiple nitpicks - spaces matter

htdat committed:
Update more specific return types

htdat commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -169,4 +192,52 @@ protected function populate_context_from_order() {
+			return null;
+		}
+
+		$this->dpps->remove_session_processing_order( $order_id );
+		$context->set_intent( $authorized_intent );
+		$context->set_detected_authorized_intent();
+
+		$new_state = $this->create_state( ProcessedState::class );
+		return $new_state->complete_processing();
```
@RadoslavGeorgiev - I am moving your question to here so it's not getting lost. 

> Question: Should we call `remove_session_processing_order` when transitioning from `ProcessedState` to `CompletedState`?

Yes, before we have already set $order_id when no duplicate order is detected.
https://github.com/Automattic/woocommerce-payments/blob/963541d30f6e78653c57325e52712fc417bdf130/src/Internal/Payment/State/InitialState.php#L210

This is also a bit more defensive even though the `clear_session_processing_order_after_landing_order_received_page` hook would also clear it when it arrives the order-received page.

RadoslavGeorgiev replied:
Sorry @htdat, could you please re-phrase this reply?  Overall, I'm not sure whether you agree we should do it, or not :D

> Yes, before we have already set $order_id when no duplicate order is detected.

Is that the `$order_id` in the session? Because we would overwrite it shortly.

htdat replied:
@RadoslavGeorgiev - Aha, sorry. My previous reply meant "yes, we should do that" and actually, we're doing that around this code (line 236). It's just that we're not doing that exactly when transitioning from `ProcessedState` to `CompletedState`? Instead, it's in the process of transiting from `InitialState` to `ProcessedState` (line 236 here). 

> Is that the `$order_id` in the session? Because we would overwrite it shortly.

Short answer is yes. 

So the process looks like: 

- Run `process_duplicate_order`. If no duplicate order is detected, return `null` as well as call `update_session_processing_order`. Code ref: 
https://github.com/Automattic/woocommerce-payments/blob/e688822fcac55cfd9bb6e66b947c376c092d355b/src/Internal/Payment/State/InitialState.php#L209-L213

- Run `process_duplicate_payment`. If the duplicate payment is detected, remove the `$order_id`, which has just been set.

RadoslavGeorgiev replied:
Ok, I think there's a gap here. Let me explain my understanding:

- The payment process starts, and in the normal flow (there is no duplicate order detected), we store the order ID in the session.
- If processing fails, we keep the order in session, and use it to prevent duplicates in the future.
- __If processing succeeds, we remove the order ID from the session.__

The last point is what we do on the Order Received page, but we should also do it at the end of the successful process, like presently do in the gateway: https://github.com/Automattic/woocommerce-payments/blob/ea5370017345fc2254f5011e4fdf381b3842e18a/includes/class-wc-payment-gateway-wcpay.php#L1501

This can only happen _after_ successful processing, but not before it, right? So in addition to what you described (`InitialState` -> `ProcessedState`), we should also clear the order from session when transition from `ProcessedState` to `CompletedState`. Am I missing something?

htdat replied:
That's correct 💯 . It was me to miss this fact when merging Zvonimir's PR that introduces the `ProcessedState`. Addressed here c7a932d0f40be7d205d4c06651b794cef06d58f0

htdat commented:
@RadoslavGeorgiev - thanks for your detailed review. I've addressed all your feedback and resolved most conversations, and keep open for two conversations that you might have other feedback.

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/7411-duplicate-prevention

htdat committed:
Changer $this->sut to $this->mocked_sut

For test_clean_up_when_detecting_duplicate_order

htdat committed:
Move back to the local var $mock_sut

htdat committed:
Move remove_session_processing_order to transition from ProcessedState to CompletedState

htdat committed:
Merge branch 'develop' into rpp/7411-duplicate-prevention

RadoslavGeorgiev reviewed (approved):
Excellent work, @htdat!

Thank you for addressing my comments! I went through manual tests as well, and everything worked like a charm.
